src/shared.c

```diff
@@ -696,248 +707,557 @@ bool write_edited_desktop_file(GKeyFile *key_file_structure, const char* appimag
-                if(verbose)
-                    fprintf(stderr, "desktop_icon_value_original: %s\n", desktop_icon_value_original);
-                if (!write_edited_desktop_file(key_file_structure, path, desktop_filename, 1, md5, verbose)) {
-                    fprintf(stderr, "Failed to install desktop file\n");
-                    return false;
-                }
 
-                g_free(desktop_filename);
+            *desktop_filename = g_strdup(g_path_get_basename(filename));
+
+            // a structure that will hold the information from the desktop file
+            *key_file = g_key_file_new();
+
+            gboolean success = g_key_file_load_from_data(*key_file, buff, size,
+                G_KEY_FILE_KEEP_COMMENTS | G_KEY_FILE_KEEP_TRANSLATIONS, NULL);
+
+            if (!success) {
+                // cleanup
+                g_free(key_file);
+                key_file = NULL;
+
+                break;
             }
-            g_key_file_free(key_file_structure);
+
+            // cleanup
+            g_free(filename);
+
+            return true;
         }
+    }
 
-        gchar *dest = NULL;
-        /* Get icon file(s) and act on them in one go */
+    g_free(filename);
 
-        if(g_str_has_prefix(filename, "usr/share/icons/") || g_str_has_prefix(filename, "usr/share/pixmaps/") || (g_str_has_prefix(filename, "usr/share/mime/") && g_str_has_suffix(filename, ".xml"))){
-            gchar *t = replace_str(filename, "usr/share", g_get_user_data_dir());
-            gchar *dest_dirname = g_path_get_dirname(t);
-            g_free(t);
+    return false;
+}
 
-            gchar *file_basename = g_path_get_basename(filename);
-            gchar *dest_basename = g_strdup_printf("%s_%s_%s", vendorprefix, md5, file_basename);
+bool archive_copy_icons_recursively_to_destination(struct archive** a, const gchar* md5,
+                                                   const gchar* desktop_icon_value_original, gboolean verbose) {
+    // iterate over all files ("entries") in the archive
+    // looking for a file with .desktop extension in the root directory
+    struct archive_entry* entry;
+    gchar* filename = NULL;
 
-            dest = g_build_path("/", dest_dirname, dest_basename, NULL);
+    bool errored = false;
+
+    int r = -1;
+
+    for (;;) {
+        r = archive_read_next_header(*a, &entry);
+
+        if (r == ARCHIVE_EOF) {
+            break;
+        }
+
+        if (r != ARCHIVE_OK) {
+            fprintf(stderr, "%s\n", archive_error_string(*a));
+
+            errored = true;
+            break;
+        }
+
+        /* Skip all but regular files; FIXME: Also handle symlinks correctly */
+        if (archive_entry_filetype(entry) != AE_IFREG)
+            continue;
+
+        filename = replace_str(archive_entry_pathname(entry), "./", "");
+
+        gchar* dest = NULL;
+
+        // Get icon file(s) and MIME types and act on them in one go
```

Comment:
What about making a function of this ?

Reply:
What _exactly_ is "this" in this context?

Reply:
What the comment says: "Get icon file(s) and MIME types and act on them in one go"
If we use a function here, the code will look clear and the function name will server as documentation.

Reply:
The idea behind this refactoring was basically to expose some functionality for use in the new function, I focused on extracting only those functions. You should have push access, feel free to extract and push.